File: backend/src/modules/auth/controllers/auth-controller.ts
import { type Request, type Response, type NextFunction } from "express";
import { registerSchema, loginSchema } from "../validators/auth-validator.js";
import * as authService from "../services/auth-service.js";
import type { AuthRequest } from "../../../middleware/auth.js";

export async function register(req: Request, res: Response, next: NextFunction) {
  try {
    const data = registerSchema.parse(req.body);
    const result = await authService.registerUser(data);

    res.cookie("accessToken", result.accessToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "lax",
      maxAge: 15 * 60 * 1000,
    });

    res.cookie("refreshToken", result.refreshToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "lax",
      maxAge: 7 * 24 * 60 * 60 * 1000,
      path: "/api/auth",
    });

    res.status(201).json({ user: result.user });
  } catch (err) {
    next(err);
  }
}

export async function login(req: Request, res: Response, next: NextFunction) {
  try {
    const data = loginSchema.parse(req.body);
    const result = await authService.loginUser(data);

    res.cookie("accessToken", result.accessToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "lax",
      maxAge: 15 * 60 * 1000,
    });

    res.cookie("refreshToken", result.refreshToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "lax",
      maxAge: 7 * 24 * 60 * 60 * 1000,
      path: "/api/auth",
    });

    res.json({ user: result.user });
  } catch (err) {
    next(err);
  }
}

export async function refresh(req: Request, res: Response, next: NextFunction) {
  try {
    const token = req.cookies?.refreshToken;
    if (!token) {
      res.status(401).json({ error: "Refresh token não fornecido" });
      return;
    }

    const result = await authService.refreshTokens(token);

    res.cookie("accessToken", result.accessToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "lax",
      maxAge: 15 * 60 * 1000,
    });

    res.cookie("refreshToken", result.refreshToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "lax",
      maxAge: 7 * 24 * 60 * 60 * 1000,
      path: "/api/auth",
    });

    res.json({ message: "Tokens renovados" });
  } catch (err) {
    next(err);
  }
}

export async function logout(req: AuthRequest, res: Response, next: NextFunction) {
  try {
    if (req.user?.id) {
      await authService.logoutUser(req.user.id);
    }

    res.clearCookie("accessToken");
    res.clearCookie("refreshToken", { path: "/api/auth" });
    res.json({ message: "Logout realizado com sucesso" });
  } catch (err) {
    next(err);
  }
}

export async function me(req: AuthRequest, res: Response, next: NextFunction) {
  try {
    const profile = await authService.getProfile(req.user!.id);
    res.json({ user: profile });
  } catch (err) {
    next(err);
  }
}

File: backend/src/modules/auth/controllers/auth_controller.py
from __future__ import annotations

from fastapi import Depends, Response

from config.settings import get_settings
from modules.auth.schemas.auth_schema import LoginInput, RegisterInput
from modules.auth.services import auth_service
from modules.auth.services.invite_service import create_invite


def _current_user():
    from middleware.auth import get_current_user
    return Depends(get_current_user)


settings = get_settings()


def _set_auth_cookies(response: Response, access_token: str, refresh_token: str) -> None:
    response.set_cookie(
        key="accessToken",
        value=access_token,
        httponly=True,
        secure=settings.is_production,
        samesite="lax",
        max_age=15 * 60,
    )
    response.set_cookie(
        key="refreshToken",
        value=refresh_token,
        httponly=True,
        secure=settings.is_production,
        samesite="lax",
        max_age=7 * 24 * 60 * 60,
        path="/api/auth",
    )


async def register(data: RegisterInput, response: Response) -> dict:
    result = await auth_service.register_user(data)
    _set_auth_cookies(response, result["accessToken"], result["refreshToken"])
    return {"user": result["user"].model_dump()}


async def login(data: LoginInput, response: Response) -> dict:
    result = await auth_service.login_user(data)
    _set_auth_cookies(response, result["accessToken"], result["refreshToken"])
    return {"user": result["user"].model_dump()}


async def refresh(response: Response, refresh_token: str | None = None) -> dict:
    if not refresh_token:
        from fastapi.exceptions import HTTPException
        raise HTTPException(status_code=401, detail="Refresh token n\u00e3o fornecido")

    result = await auth_service.refresh_tokens(refresh_token)
    _set_auth_cookies(response, result["accessToken"], result["refreshToken"])
    return {"message": "Tokens renovados"}


async def _logout(response: Response, user_id: str) -> dict:
    await auth_service.logout_user(user_id)
    response.delete_cookie("accessToken")
    response.delete_cookie("refreshToken", path="/api/auth")
    return {"message": "Logout realizado com sucesso"}


async def _get_profile(user_id: str) -> dict:
    profile = await auth_service.get_profile(user_id)
    return {"user": profile.model_dump()}


async def _update_profile(user_id: str, data: dict) -> dict:
    profile = await auth_service.update_profile(user_id, data)
    return {"user": profile.model_dump()}


async def _generate_invite(user_id: str) -> dict:
    result = await create_invite()
    return {"invite": result}
